### ISSUE TITLE
Calculation for Leap Years is inaccurate

### ISSUE THREAD

**Issue by Generalcamo** [OpenLoco/OpenLoco#1671]
**Version information:**
 - Platform: Windows 10
 - Version: 22.09
 - Build/commit: 00b883a

**Describe the bug**
February 29th follows from February 28th on the years 2100, 1900 and 1800. These years are not leap years, and therefore these dates do not actually exist.

**To Reproduce**
Steps to reproduce the behaviour:
1. Start (or create) a scenario beginning in either 1800, 1900, or 2100
2. Wait for the date to arrive at February 28th
3. The next date will be February 29th

**Expected behaviour**
These dates do not exist in the Gregorian Calculator, and these years should not be treated as a leap year.

**Additional context**
This is an original bug.

**Comment by LeftofZen** [comment]
I've known about this for a while but I just ignored it because I never considered it a bug; just a simplification to dates made by CS. I'm happy to make this change one day, but wouldn't cause a sizeable deviation from the vanilla simulation? I always thought we were aiming to keep OpenLoco as consistent with vanilla as possible?

**Comment by AaronVanGeffen** [comment]
Yeah, I brought it up when I implemented `Scenario::initialiseDate`, too: https://github.com/OpenLoco/OpenLoco/pull/791#discussion_r581881329. We deliberately didn't change the logic at the time. Good to have it tracked in an issue, though.

**Comment by Generalcamo** [comment]
> I always thought we were aiming to keep OpenLoco as consistent with vanilla as possible?

It's been stated with the sister project, OpenRCT2, that this isn't exactly the case (see: https://github.com/OpenRCT2/OpenRCT2/issues/2382). I would imagine that a similar instance applies here, but I could be incorrect.

**Comment by LeftofZen** [comment]
@Generalcamo the reverse-engineering has already been done; no more is needed to implement this fix - feel free to submit a PR!
https://github.com/OpenLoco/OpenLoco/blob/14eb1eb5c656b7ef74f28417111024c5021e925c/src/OpenLoco/Date.cpp#L15-L18

I personally don't have an issue with this fix, I like it, but I'll defer the ruling on this to the other more experienced team members.

**Comment by Generalcamo** [comment]
I saw that and I have an easy fix. But if there are other areas in the code that make assumptions on this, then we won't know. Better to wait.

**Comment by auge8472** [comment]
@LeftofZen Your fix is incomplete, when it should follow the real rules of the Gregorian calendar. There is an exception from the four years rule for "secular" years (like 1800, 1900 and 2100) with the exception from this exception for "secular" years, that are dividable by 400 without a rest (like 2000 or 2400).

I don't know if this is of any relevance for the game or if it would break something. But in general I also don't think, that the leap years are of any relevance for the game at all. If the game years counts 365 days or every fourth 366 (with or without the exceptions) IMHO makes no real difference for the game playing.

**Comment by LeftofZen** [comment]
@auge8472 I'm not sure what you mean, I didn't provide any fix? I just highlighted the lines of code that currently exist for leap year calculation

**Comment by Generalcamo** [comment]
Here's the fix I propose:
`return year % 4 == 0 && (year % 100 != 0 || year % 400 == 0)`

**Comment by LeftofZen** [comment]
If we're re-visiting this, I'd like to know how much this would cause openloco to deviate from vanilla - how much will it affect the replays we have that we use to compare between?

**Comment by duncanspumpkin** [comment]
I would rather not change it at this time.
